### ISSUE TITLE
Work still to be done with the heterogeneous agents methods

### ISSUE THREAD

**Issue by jagman88** [EconForge/dolo#112]
I recently added a file (with examples) designed to solve heterogeneous agents models with no aggregate uncertainty using the dtcscc methods in dolo (see [PR #111](https://github.com/EconForge/dolo/pull/111). 

This project is still a work in progress, and there is more yet to be done. Here is a quick list of things that I can see still need to be done (@albop, @spencerlyon2, you may also have ideas):

- Write tests for the various funtions inside `heterogeneous_agents.py`

- Modify the function `solve_eqm` to solve for equilibrium in more than one aggregate state variable.

- Modify the function `aggregate_resid` to take non-straightforward aggregate residual functions. To take a straightforward example (currently supported): the aggregate residual in the incomplete markets model only involves the aggregate capital demand minus the aggregated-up capital supply. A non-straightfoward example (not currently supported): the residual from aggregate consumption computed using a Dixit-Stiglitz aggregator (i..e. involves a integral of a non-linear function of the distributon of consumption types).

- In the function `stat_dist`, the distribution's transition matrix is constructed using row-wise Kronecker products. The component transition matrices (Qm and Qs) are initially sparse matrices, but are currently taken back to arrays for the computation of the Kronecker. @albop mentioned that he has a routine to do this, keeping the matrices in sparse format. 

- Currently, the number of endogenous variables and exogenous variables in the model are inferred from the number of shocks and the number of total states. This is refererenced with "hack" notes throughout the code. This information can likely be gleaned from the model in some other less hack-y way. 

- The code currently does not support i.i.d shocks to non-AR(1) model equations. It would be possible to incorporate these, however it would involve a reasonably significant re-writing of the `stat_dist` function and associated code.

- The `exog_grid_trans` function makes the strong assumption that AR(1) processes in the model are independent of each other. To test this it checks that the covariance matrix from the yaml file is diagonal, and throws an exception if it is not. Rather than this exception, it would be good to pass off the discretization of the AR(1)s in this case to the `dolo.numeric.discretization.multidimensional_discretization` function. There, the assumption is that the AR(1)s are not indpendent, but that their persistence parameters are identical.

**Comment by duzaichuan** [comment]
Excited about your great work on HACK! It really makes the modelling easy to be done.

**Comment by albop** [comment]
Thanks @duzaichuan. We take all compliments. But are you really referring to dolo (where we are still at the experimentation phase with hetag), or to https://github.com/econ-ark/HARK ?

**Comment by duzaichuan** [comment]
Yes! Sorry, I should say dolo. HARK is an excellent project, too. But I prefer dolo's simplicity indeed. Looking forward to the official release~

**Comment by albop** [comment]
Thanks! Upcoming...
